### PR TITLE
commit and push a container proposal

### PR DESCRIPTION
**Pull request by mqliang**
previous discussion:
https://github.com/kubernetes/kubernetes/pull/19919
https://github.com/kubernetes/kubernetes/issues/14561
https://github.com/kubernetes/kubernetes/pull/19107

Our team has implement a POC kubepush agent, see https://github.com/mqliang/kubepush. We'd like contribute it to community(contrib repo seems a good place) if it looks good.

@bgrant0607 @smarterclayton @derekwaynecarr @vishh

cc @ddysher

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34384)

### CODE REVIEW

**Comment by mqliang** [comment]
Our team has implement the kubepush agent, see https://github.com/mqliang/kubepush. We'd like contribute it to community(contrib repo seems good place) if it looks good.

**Comment by zhouhaibing089** [comment]
Could you help me understand the motivation more clearly? say what are the cases that a container change must be made in k8s while it could not be a simple local commit and then push to registry?

File: docs/proposals/push.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  namespace: ns1
+  name: name1
+  labels:
+    kubepush.alpha.kubernetes.io/nodename: i-jgganq
+spec:
+  podName: master-nginx-i-94uugzhjm
+  containerName: master-nginx
+  image: cargo.caicloud.io/liangmq/nginx: 1.2
+  imagePushSecrets:
+  - name: cargo.caicloud.io
```
Comment (zhouhaibing089) [inline docs/proposals/push.md]: How to create a push secrets, the same way as creating pull secrets?

File: docs/proposals/push.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+Note:
+
+* Why we need a `kubepush.alpha.kubernetes.io/nodename` label: this label
+  indicate which node the container user whant commit and push is running on
+
+
+######3. kubepush daemonset
+There is a kubepush daemonst running on every node, list&watch `Push` resource
+and
```
Comment (zhouhaibing089) [inline docs/proposals/push.md]: When will the daemonset decide to commit and push to registry? I am feeling more descriptions are needed.

File: docs/proposals/push.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+metadata:
+  name: push.k8s.io
+apiVersion: extensions/v1beta1
+kind: ThirdPartyResource
+description: "Allow user to commit a container and push it into registry"
+versions:
+- name: v1
+```
+
+###### 2. create a `Push` resource
```
Comment (zhouhaibing089) [inline docs/proposals/push.md]: The name Push feels more like a verb, is there any other options that you've considered?

File: docs/proposals/push.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+container and push it into registry. Example `Push` resource is as follows:
+
+```
+kind: Push
+apiVersion: k8s.io/v1
+metadata:
+  namespace: ns1
+  name: name1
+  labels:
+    kubepush.alpha.kubernetes.io/nodename: i-jgganq
```
Comment (zhouhaibing089) [inline docs/proposals/push.md]: do I have to fill the value? once we know the pod name, the node name should already be a determined one..

File: docs/proposals/push.md (hunk 1 above)
Comment (mqliang) [inline docs/proposals/push.md]: yes

File: docs/proposals/push.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+spec:
+  podName: master-nginx-i-94uugzhjm
+  containerName: master-nginx
```
Comment (mqliang) [inline docs/proposals/push.md]: if you want, you definitely could.

**Comment by mqliang** [comment]
@zhouhaibing089 
>  say what are the cases that a container change must be made in k8s while it could not be a simple local commit and then push to registry?

You can definitely make a local commit and push to registry. But with k8s supporting, life will be more convenient: for example, user have a pc without docker installed, or they can not access registry from their pc(for example, registry is private and have no external IP), but they can access k8s API server, so then could `kubectl exec -it`(from web terminal) into a container. Then, user `kubectl exec -it` into a container and make some change(i.e. install some package), he want commit and push the container into registry.

File: docs/proposals/push.md (hunk 3 above)
Comment (mqliang) [inline docs/proposals/push.md]: Do you have any idea?

**Comment by brendandburns** [comment]
I think this needs to go through the kubernetes incubation process:

https://github.com/kubernetes/community/blob/master/incubator.md

So I'm going to close this proposal, if you want this to be github.com/kubernetes/kubepush let's work through incubation.

Thanks!

**Pull request closed without merging**